cpu/nrf51822: added minimal radio driver implementation

haukepetersen opened the pull request:
The NRF51822 is primarily build for BLE, but the radio also supports some other (proprietary) modes of operation. This driver uses the radio in a IEEE802.15.4 like fashion - it supports short addresses, PAN Ids, a broadcast address (0xffff) and some well known radio stuff (channels, tx power adjustment).

EDIT
~~The driver has full support for the new network stack (e.g. `ng_netdev`), and could be used as reference implementation for register-mapped network devices.~~
There are some new insights on clean driver development that are being discussed at the moment, so forget about the reference implementation part!

LudwigOrtmann commented:
=)

haukepetersen commented:
@LudwigOrtmann: just assign someone else if you cannot test this, just figured that I did most of the testing with your airfy-boards and your new position :-)

LudwigOrtmann commented:
I am unfamiliar with all the details of the new network stack, so I can't review that part. I will be happy to test the network driver during the weekend if I find time.

haukepetersen commented:
@authmillenon, so that leaves you to look over the code?!

haukepetersen commented:
@LudwigOrtmann thx for testing, though!

authmillenon commented:
I'll do that if I catch up my lost sleep from last night :sleepy:.

LudwigOrtmann commented:
When building (rebased on master) with `CFLAGS=-DDEVELHELP`, the thread named `dump` seems to have memory problems:
```
2015-03-26 21:35:58,520 - INFO # > ps
2015-03-26 21:35:58,521 - INFO # 
2015-03-26 21:35:58,528 - INFO # 	pid | name                 | state    Q | pri | stack ( used) | location   
2015-03-26 21:35:58,536 - INFO # 	  1 | idle                 | pending  Q |  15 |   256 (  140) | 0x20000170 
2015-03-26 21:35:58,543 - INFO # 	  2 | main                 | running  Q |   7 |  1536 (  724) | 0x20000270 
2015-03-26 21:35:58,550 - INFO # 	  6 | uart0                | bl rx    _ |   6 |  1024 (  240) | 0x200009a0 
2015-03-26 21:35:58,558 - INFO # 	  4 | nomac                | bl rx    _ |   5 |  1024 (  268) | 0x200013a0 
2015-03-26 21:35:58,566 - INFO # 	  5 | dump                 | bl rx    _ |   6 |  1536 ( 1536) | 0x20000da0 
2015-03-26 21:35:58,573 - INFO # 	  6 | uart0                | bl rx    _ |   6 |  1024 (  240) | 0x200009a0 
2015-03-26 21:35:58,579 - INFO # 	    | SUM                  |            |     |  6400 ( 3148)
```

OlegHahm commented:
Keep in mind that this kind of ps output often indicates that something else than the thread itself overflows.

OlegHahm commented:
Looking at the memory map may help.

LudwigOrtmann commented:
BTW: a first functionality test was positive (no false positives due to addressing, broadcast works, addressed communication works).
```
2015-03-27 09:00:57,189 - INFO # PKTDUMP: data received:
2015-03-27 09:00:57,195 - INFO # ~~ SNIP  0 - size:   3 byte, type: NETTYPE_UNDEF (0)
2015-03-27 09:00:57,201 - INFO # ~~ SNIP  1 - size:  10 byte, type: NETTYPE_UNDEF (0)
2015-03-27 09:00:57,206 - INFO # ~~ PKT    -  2 snips, total size:  13 byte
```

haukepetersen commented:
No idea yet, where the full stack might come from. But I am pretty confident that it is unrelated to this PR, would you agree?

LudwigOrtmann commented:
Can't say, didn't debug.

LudwigOrtmann commented:
What I'm missing is in the documentation is information on how this driver uses the radio.
I.e.: What PHY/LLC/MAC is used, how are they adopted for providing 802.15.4 functionality, what are the implications?

LudwigOrtmann commented on cpu/nrf51822/include/nrfmin.h:
```diff
@@ -0,0 +1,81 @@
+
+/**
+ * @defgroup        cpu_nrf51822_nrfmin NRF Minimal Radio Driver
+ * @ingroup         cpu_nrf51822
```
This should be `drivers`.

LudwigOrtmann replied:
This should be `drivers`.

haukepetersen replied:
no, this driver is part of the CPU

LudwigOrtmann replied:
I know, but for consistency I would put it in drivers because that is where I would look for such a driver in the documentation. Other driver implementations are part of CPU as well (e.g. every peripheral driver).

haukepetersen replied:
fine :-)

authmillenon commented on cpu/nrf51822/radio/nrfmin/nrfmin.c:
```diff
@@ -0,0 +1,670 @@
+    _tx_buf.src_addr[1] = (uint8_t)(_addr);
+    _tx_buf.dst_addr[0] = dst_addr[0];
+    _tx_buf.dst_addr[1] = dst_addr[1];
+    _tx_buf.proto = (uint16_t)payload->type;
```
Since this the values of `nettype_t` are dependent on the module make-up during compile time I would highly recommend to use the ethertype (also 16-bit long) or similar for `proto`. Otherwise you are basically putting the constraint on this driver that it only works with an equally compiled network stack in RIOT.

haukepetersen replied:
this is the case anyhow :-)

authmillenon replied:
No... You can even let e.g. Linux and Windows speak to each over WiFi other because Ethertype is universal and not dependent on what is given at compile time ;)

authmillenon replied:
All I'm saying is: please don't set a bad example and send `nettype_t` over the air. This is not what the type is intended for and nobody should ever do that. A protocol (even if it is just invented by yourself) needs to have some consistency and not being dependent on a certain build.

haukepetersen replied:
you are right, will be fixed.

authmillenon commented on cpu/nrf51822/radio/nrfmin/nrfmin.c:
```diff
@@ -0,0 +1,670 @@
+    ng_netif_hdr_set_dst_addr(hdr, data->dst_addr, 2);
+
+    /* allocate and fill payload */
+    pkt = ng_pktbuf_add(pkt_head, data->payload, data->length - 6, data->proto);
```
See https://github.com/RIOT-OS/RIOT/pull/2711/files#r27354332

haukepetersen replied:
will fix it.

authmillenon commented on cpu/nrf51822/radio/nrfmin/nrfmin.c:
```diff
@@ -0,0 +1,670 @@
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t length;                     /**< packet length */
+    uint8_t src_addr[2];                /**< source address of the packet */
```
Constant value vs. macro.

haukepetersen replied:
This I will not change. The semantics are clear and a 2 is faster to read/grasp then another define...

haukepetersen commented:
found the problem with the stack size usage shown with the `ps` shell command: the `pktdump` thread is started without the `CREATE_STACKTEST` option...

LudwigOrtmann commented:
Please add this insight to the known problems/faq/whatever section in the wiki.

haukepetersen commented:
How about I just fix it for the `pktdump` module?

haukepetersen commented:
how about I just fix it: #2758

LudwigOrtmann commented:
I did not suggest to add this entry for this particular module, but for the underlying problem, i.e. the missing flag.

LudwigOrtmann commented:
@haukepetersen please rebase.

haukepetersen commented:
rebased

haukepetersen commented:
introduced some means for globally distinctive payload type identifiers for this driver.

haukepetersen commented:
changed doxygen group and rebased

LudwigOrtmann commented on cpu/nrf51822/include/nrfmin.h:
```diff
@@ -0,0 +1,81 @@
+ *   - setting of channel (0 to 0x3f)
+ *   - setting of TX power (+4dBm to -20dBm)
+ *   - packet type labeling
+ *   - setting device state (RX, SLEEP)
```
This documentation should be part of the interface, not the file.

LudwigOrtmann commented:
> What I'm missing in the documentation is information on how this driver uses the radio.
I.e.: What PHY/LLC/MAC is used, how are they adopted for providing 802.15.4 functionality, what are the implications?

Still missing it.

LudwigOrtmann commented:
@haukepetersen ping

haukepetersen commented:
In my opinion the documentation is quite sufficient and clear, but I can add some more explanations. I just don't know when I get to it...

LudwigOrtmann commented:
The relevant documentation as of now is:
```diff
+ * This driver enables the use of the NRF51822 radio in a 802.15.4 like fashion.
+ * The driver does however not support the 802.15.4 frame format, but uses its
+ * own packet format.
```
It does not explain *how* it does this. As an application developer, I want to know if every device wakes up when a packet is transmitted or only the one with the correct MAC address, ...

haukepetersen commented:
read the Nordic manual! But as I said, I will add some more text - once I find some time.

d00616 commented:
@haukepetersen Thank you for this driver. It looks like its working.

@LudwigOrtmann
 > It does not explain how it does this. As an application developer, I want to know if every device wakes up when a packet is transmitted or only the one with the correct MAC address, ...

An interrupt is only triggered, when device or broadcast address matches but rx is on most time. This needs 13.4mA@2Mbps. An full running CPU core needs 4.1mA when code is executed from flash.

LudwigOrtmann commented:
@d00616 thank's but this needs to go into the driver documentation ;)

d00616 commented on cpu/nrf51822/radio/nrfmin/nrfmin.c:
```diff
@@ -0,0 +1,730 @@
+#define CONF_S1                 (0U)
+#define CONF_STATLEN            (0U)
+#define CONF_BASE_ADDR_LEN      (4U)
+#define CONF_ENDIAN             RADIO_PCNF1_ENDIAN_Big
```
In "Reference manual v3.0 17.1.13" is documented using big endian disables "Device Address Match" feature. This feature looks useful for addressing.

OlegHahm commented:
> @d00616 thanks but this needs to go into the driver documentation ;)

I would say that's a nice-to-have, but not a must. Hence, if this documentation is the only missing piece here, I would strongly vote for merge this ASAP and improve the documentation later if needed.

LudwigOrtmann commented:
It's not that hard to copy-paste one sentence, besides the documentation has another issue as pointed out above.

d00616 commented on cpu/nrf51822/include/nrfmin.h:
```diff
@@ -0,0 +1,81 @@
+/**
+ * @brief   Default channel set after initialization
+ */
+#define NRFMIN_DEFAULT_CHANNEL      (1U)                /* 2401MHz */
```
Please set it to >=4. Bluetooth is starting at channel 4 to take sure there is no interference with frequencies less than 2400MHz.

I prefer channel 49 or 50. This frequency is outside of all WIFI channels and usable world wide. -> http://en.wikipedia.org/wiki/List_of_WLAN_channels

haukepetersen replied:
> Bluetooth is starting at channel 4

One more reason to set it to something lower than that. Remember, this driver is not meant to be used with a Bluetooth stack, it just provides a very simplified usage of the radio device in a 802.15.4 like scheme...

d00616 replied:
@haukepetersen Please look into the radio characteristics nRF51822_PS_v3.1.pdf

Chapter: 8.5.1 +/-320KHz typically frequency deviation at 2Mbps without min or max values, With BLE maximal +/-275KHz

Chapter 8.5.5 RX selectivity - modulated interfering signal
@2Mbps: 12db at channel, -4db at channel+-2MHz, -24db at channel+-4MHz, -28db at channel+-6MHz, -44db at channel+-12MHz
@BLE: 4db at channel, -10db at channel+-1MHz, -25db at channel+-2MHz 

I think both characteristics are reasons to use another channel. 2MBs uses a lot more bandwith then BLE modulation. Maybe it's safe to use this channel @250kbps but i don't know it. I'm sure there are to much submissions outside the ISM range with this configuration.

haukepetersen commented:
@LudwigOrtmann improved the documentation, is this what you were looking for?

haukepetersen commented:
rebased

LudwigOrtmann commented:
@haukepetersen great!

haukepetersen commented:
should I squash?

LudwigOrtmann commented:
yes please

d00616 commented on cpu/nrf51822/include/nrfmin.h:
```diff
@@ -0,0 +1,102 @@
+ * with other NRF51822 devices using the same driver.
+ *
+ * The driver is using a Nordic proprietary physical layer, configured to for a
+ * bitrate of 2Mbit. The payload length is set to a maximum length of 250 byte.
```
Sorry for my note here but 250 bytes are to much, when an 60ppm crystal is used. In reference manual v3.0 "17.1.9 Maximum consecutive transmission time" is to read that the radio trasmission time is to limit up to 1ms with an 60ppm crystal and to 16ms with an 30ppm crystal. In 250ms its possible to transmit 250 bytes including preamble, addresses and CRC.
I think it's an good idea to reduce this value or add a note which crystal has to use here.

haukepetersen replied:
Good point, but I say let's not get too academic here... I tested with cheap China modules and 250 byte packets, and that worked fine... So in my opinion its fine to go a little out of spec here...

haukepetersen commented:
And can anybody tell my why Travis did not run for over **3** days?

OlegHahm commented:
I have no idea. I don't get Travis' queue management... But killing outdated builds on the Travis page manually helps a little bit.

LudwigOrtmann commented:
@haukepetersen Travis ran and disliked your trailing whitespaces.

haukepetersen commented:
hm, for some reasons one file had windows line endings. As I am only using Linux computers, I always wonder where they come from...

Anyway, fixed, squashed and rebased, can we merge?

d00616 commented:
@haukepetersen cpu/nrf51822/include/nrfmin.h is empty

haukepetersen commented:
Fell of the wagon while squashing, should be fixed now.

LudwigOrtmann commented on cpu/nrf51822/include/nrfmin.h:
```diff
@@ -0,0 +1,102 @@
+ *    byte0   byte1   byte2   byte3   byte4   byte5   byte6   byte7   byteN
+ * --------------------------------------------------------------------------
+ * | length |   src_addr    |   dst_addr    |    proto      |   payload...  |
+ * --------------------------------------------------------------------------
```
Doxygen interprets these hyphens as a headline marker... maybe a table?
```diff
- *    byte0   byte1   byte2   byte3   byte4   byte5   byte6   byte7   byteN
- * --------------------------------------------------------------------------
- * | length |   src_addr    |   dst_addr    |    proto      |   payload...  |
- * --------------------------------------------------------------------------
+ *   byte0 | byte1 - byte2 | byte3 - byte4 | byte5 - byte6 | byte7 - byteN
+ *  -------|---------------|---------------|---------------|--------------
+ *  length |   src_addr    |   dst_addr    |    proto      |   payload...
```

d00616 commented:
If we want to ignore that using 2MBit/s on channel 1 is possible illegal, this can be merged after fixing the doxygen error.

haukepetersen commented:
fixed doxygen, squashed and rebased. Lets wait for Travis and merge.

haukepetersen commented:
Travis is happy -> go.

OlegHahm commented:
Where was the ACK?

haukepetersen commented:
it was implicitly stated above... :)
